### DISCUSSION TITLE
to many tools ??

### DISCUSSION THREAD

**Discussion by michaelleewebb** [twinbasic/twinbasic#1733] [Q&A] [answered]
not everytime that i change the code and recompile, but nearly everytime, i get a new menu item on my vbide form. see attached photo.
![to many tools](https://github.com/twinbasic/twinbasic/assets/7157330/88f3d406-b776-4699-ae87-69ee1bbe40eb)
note the "mike tools" menu item over and over.
in the vb ide under tools/references i see myCOMaddin but it is unchecked. 
on the command line i tried:
regsvr32 -u C:\twinbasic\comAddin\Build\mycomaddin_win32.dll
this works so far as unregisting the dll, (the first time) but still i have all the "mike tools" on the menu of the vb IDE. 

inside the twinBasic ide i have tried the clean (deregister & delete build) icon. also the restart the compiler icon. no luck.

if i create a brand new database program the tools menu items are there before i do anything else.

anyone know how to get rid of them.

**Comment by GCuser99** [comment]
I don't think it is a registration problem - you probably have "stranded" toolbar menus because your Add-in is not properly removing the menu on Disconnect/Shutdown event due to a coding error. Once the menu gets stranded, then you have to correct your error so you do not create any more stranded menus, and then you need remove the stranded ones. 

Can you post the MyComAddin class code that created "Mikes Tools"? Let's first make sure that you are correctly removing the menu in your current Add-in, and then we can address how to remove the stranded menus.

**Comment by michaelleewebb** [comment]
here is code, i believe you are correct`

Class MyCOMAddin
 
    ' ---------------------------------------------------------------------------------------
   ' For further instructions, please visit https://www.twinbasic.com/preview.html#sample5
   ' ---------------------------------------------------------------------------------------
    
   'Note from GCuser99:
   'this requires a reference to Microsoft Office XX.0 Object Model
   'and the Microsoft Visual Basic for Applications Extensibility

   Implements IDTExtensibility2

   /*
       PLEASE NOTE: make sure you create a DLL of the correct bitness to match your version of VBA
       (e.g. if you're using the 64-bit version of VBA, make sure you change the 'Active Build' in the twinBASIC panel to 'win64')
   */
    
   /*
   [ WithDispatchForwarding ]
   Implements IRibbonExtensibility
   */
    
   /* 
       The [WithDispatchForwarding] attribute seen above is needed so that late-bound calls
       on the IRibbonExtensibility interface get routed to our MyCOMAddin default interface.
       Without it, events like OnHelloWorldClicked (defined below) will not fire.
   */
    
   Private applicationObject As Object ' Excel.Application, late-bound, but you can add a project reference if you wish
    
   Private vbe As VBIDE.VBE
   Private WithEvents menuItem1Events As VBIDE.CommandBarEvents
   Private isConnected As Boolean

   /* 
       Firstly we implement the IDTExtensibility2 interface members
       We don't do anything particularly interesting here, except for taking a reference
       to the Application object provided to us in the OnConnection event
   */
   Sub OnConnection(ByVal Application As Object, _
                           ByVal ConnectMode As ext_ConnectMode, _
                           ByVal AddInInst As Object, _
                           ByRef custom As Variant()) _
       Implements IDTExtensibility2.OnConnection
        
       'MsgBox "MyCOMAddin.OnConnection"
       Set applicationObject = Application
        
       Set vbe = Application.vbe
       isConnected = True
       CreateVBEMenu()
   End Sub
    
   Sub OnDisconnection(ByVal RemoveMode As ext_DisconnectMode, _
                       ByRef custom As Variant()) _
       Implements IDTExtensibility2.OnDisconnection
       ShutdownAddin
       'MsgBox "MyCOMAddin.OnDisconnection"
   End Sub
    
   Sub OnAddInsUpdate(ByRef custom As Variant()) _
       Implements IDTExtensibility2.OnAddInsUpdate
        
       'MsgBox "MyCOMAddin.OnAddInsUpdate"
   End Sub
    
   Sub OnStartupComplete(ByRef custom As Variant()) _
       Implements IDTExtensibility2.OnStartupComplete
        
       'MsgBox "MyCOMAddin.OnStartupComplete"
   End Sub
    
   Sub OnBeginShutdown(ByRef custom As Variant()) _
       Implements IDTExtensibility2.OnBeginShutdown
       ShutdownAddin
       'MsgBox "MyCOMAddin.OnBeginShutdown"
   End Sub
    
   /* 
       Next we implement IRibbonExtensibility.GetCustomUI which lets us provide
       a ribbon XML string to the host
   */
    
   /*
   Private Function GetCustomUI(ByVal RibbonID As String) As String _
       Implements IRibbonExtensibility.GetCustomUI

       Dim strXML As String
       strXML &= "<customUI xmlns=""http://schemas.microsoft.com/office/2006/01/customui"">" & vbCrLf
       strXML &= "  <ribbon startFromScratch=""false"">" & vbCrLf
       strXML &= "    <tabs>" & vbCrLf
       strXML &= "      <tab id=""tabTest"" label=""twinBASIC Test"">" & vbCrLf
       strXML &= "        <group id=""grpTwin1"" label=""Hello Group"">" & vbCrLf
       strXML &= "          <button id=""btnHello"" size=""large"" label=""Hello World"" getImage=""OnGetHelloWorldButtonImage"" onAction=""OnHelloWorldClicked""/>" & vbCrLf
       strXML &= "        </group>" & vbCrLf
       strXML &= "      </tab>" & vbCrLf
       strXML &= "    </tabs>" & vbCrLf
       strXML &= "  </ribbon>" & vbCrLf
       strXML &= "</customUI>" & vbCrLf
        
       Return strXML
   End Function
   */
    
   /* 
       The XML we provided for the ribbon has a button with an onAction property of 'OnHelloWorldClicked'
       When clicked, that will trigger our event...
   */
    
   /*
   Public Sub OnHelloWorldClicked(Control As IRibbonControl)
       MsgBox "Hello world, from twinBASIC!" & vbCrLf & _
                   "(via " & applicationObject.Name & ")"
        
       ' Here we can use applicationObject (Excel.Application object, or Access.Application etc) 
       '  to interact with the host instance.
   End Sub
   */
    
   /* 
       The XML we provided for the ribbon has a button with a getImage property of 'OnGetHelloWorldButtonImage'
       The host will call this routine to obtain an image to display for the ribbon button...
       We have included twinBASIC_icon.bmp (a 24-bit bitmap image) in the Resources\BITMAP\ folder which you can see in the VS code project folder explorer on the left
   */
    
   /*
   Public Function OnGetHelloWorldButtonImage(Control As IRibbonControl) As IPictureDisp
       Return LoadResPicture("twinBASIC_icon.bmp", vbResBitmap)
   End Function
   */
    
   Private Sub menuItem1Events_Click(ByVal CommandBarControl As Object, _
               ByRef Handled As Boolean, ByRef CancelDefault As Boolean) Handles menuItem1Events.Click
       'this is where you could call your procedure with Application.Run....
       MsgBox "Hello world, from twinBASIC!" & vbCrLf & _
                         "(via " & applicationObject.Name & ")"
       Handled = True
       CancelDefault = False
   End Sub
    
   Private Sub CreateVBEMenu()
       Dim menu As CommandBarPopup
       Dim menuItem1 As CommandBarButton
        
       Set menu = vbe.CommandBars("Menu Bar").Controls.Add(Type:=msoControlPopup)
       With menu
           menu.Caption = "twin&BASIC Tools"
           menu.Tag = "twinBASICTools"
           Set menuItem1 = .Controls.Add(Type:=msoControlButton)
           With menuItem1
               .Caption = "&Hello World"
               Set menuItem1Events = vbe.Events.CommandBarEvents(menuItem1)
           End With
       End With
       Set menuItem1 = Nothing
       Set menu = Nothing
   End Sub
    
   Private Sub RemoveVBEMenu()
       /*
       Dim ctl As Object
       Do
           Set ctl = vbe.CommandBars("Menu Bar").FindControl(Tag:="twinBASICTools")
           If ctl IsNot Nothing Then
               ctl.Delete
           End If
       Loop Until ctl Is Nothing
       */
       'below can replace above loop, after debugging phase is complete
       vbe.CommandBars("Menu Bar").Controls("twinBASIC Tools").Delete
   End Sub
    
   Private Sub ShutdownAddin()
       If isConnected = False Then Exit Sub
       RemoveVBEMenu
       Set vbe = Nothing
       Set applicationObject = Nothing
       isConnected = False
   End Sub
     
End Class`

**Comment by michaelleewebb** [comment]
hmm, let me know if you can see that copy and paste i just did.

**Comment by GCuser99** [comment]
@michaelleewebb, the way you paste in code is to use three accent keys (top left on my laptop, just below the esc key) + "vba" to open the code section, and then three to close, like this:
````
```vba
my code
```
````
Is that the unaltered version of the code I gave you? It worked perfectly for me. Is that version working for you? Can you share the version that created "Mikes Tools"?

**Reply by michaelleewebb** [reply]
thats right, my mistake. i took the code you gave me and started making changes. when i got myself into trouble i recopied you code over what i had written. then i tried again and i had yet another menu item, but this menu item had reverted back to saying twin basic tools instead of mikes tools. i've copied over what i had written earlier so i can't show you what i did wrong. can we more forward with the " "stranded" toolbar menus" thing. if i know how to do that when i make a mistake (again), i'll know how to fix it.

**Comment by michaelleewebb** [comment]
i think i have it. i went to customize toolbar. then select menu, then press reset.

**Comment by GCuser99** [comment, marked as answer]
@michaelleewebb, good to hear...

COM can be difficult to debug. Here are some things you can do to help keep out of hot water...

Develop/test the non-COM part of your code in the host application (Access VBA in your case) as much as possible before moving it to the tB COM project. In your host application you can more easily/quickly pinpoint and correct logic errors. Remember that tB is backward compatible with VBA so if it runs in VBA, it should run in tB.

If there is something you don't know how to do, search the online VBx forums and the [tB Wiki](https://github.com/twinbasic/documentation/wiki) before posting here. Keep in mind that tB is not a general VBx programming forum - discussion posts here ideally should be more specifically about tB-related issues and language features. I recently joined Stack Overflow and VB Forums to learn more about how to do things with VB6 which I can then apply to my tB projects.

For the tB COM project part, before you change anything from the sample, make sure it runs correctly in your host document. In this particular case, that might include running the "Hello World" menu item. Closing the document, then reopening, rerunning "Hello World", unloading/reloading the Add-in via Options->Addins, etc. to make sure the Add-in project works on your system. If it doesn't work, then we know it is not something you did, but rather a bug or incompatibility with the code you started with.

Learn as much as you can about how the sample COM code works before changing it. You will notice that Wayne put some commented-out MsgBox's in each of the event procedures. Uncomment these and recompile/rerun to see the progression of events as you run the sample in your host application. 

Be careful about (re)registration (see [DLL hell](https://en.wikipedia.org/wiki/DLL_Hell)!). If/when you want to change the name of the Add-in from MyCOMAddin to something more meaningful to you, first unregister the original Add-in version. The way you change the Add-in name in tB is via Project->Project Properties->Project Name. You do not have to (nor should you unless you really know what you are doing) change the class name in the code pane - just in Project Properties.

Adopt some sort of project version control system for yourself, saving intermediate project versions that work without error, so that you can "revert" back if you make changes that result in a problem.

All of that should make programming life easier... Good luck!

EDIT: see [How to Create a COM Add-in](https://nolongerset.com/create-com-addin-with-twinbasic/) by Mike Wolfe

**Comment by michaelleewebb** [comment]
i will keep all of that in mind. normally i'm pretty methodical about my coding. this slipped by me. i do wish you had a "twin basic users forum" though. i will mark this as solved.

**Reply by WaynePhillipsEA** [reply]
Our discord server is our less formal "forum": #629 Feel free to join us anytime.

**Comment by GCuser99** [comment]
Have you tried [VB Forums](https://www.vbforums.com/showthread.php?890181-TwinBasic)? Everything VBA, VB6, and twinBASIC!
